### PR TITLE
Add `VnEntropy` MP class

### PR DESCRIPTION
**Pull request by AlbertMitjans**

### CODE REVIEW

File: pennylane/measurements/vn_entropy.py [hunk 1]
```diff
@@ -61,4 +64,13 @@ def circuit_entropy(x):
...
-    return MeasurementProcess(VnEntropy, wires=wires, log_base=log_base)
+    return _VnEntropy(VnEntropy, wires=wires, log_base=log_base)
+
+
+class _VnEntropy(StateMeasurement):
+    """Measurement process that returns the Von Neumann entropy."""
+
+    def process_state(self, state: np.ndarray, device_wires: Wires):
+        return qml.math.vn_entropy(
+            state, indices=self.wires, c_dtype=np.complex128, base=self.log_base
```
Comment (Jaybsoni) [inline pennylane/measurements/vn_entropy.py:75]: Similar concern as with the `mutual_info` PR. I worry that we might be forcing an interface by setting the dtype here.

File: tests/measurements/test_vn_entropy.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for the vn_entropy module"""
+import numpy as np
+
+import pennylane as qml
+
+
+class TestVnEntropy:
```
Comment (Jaybsoni) [inline tests/measurements/test_vn_entropy.py:20]: Probably worth having a more complicated test with non-zero Von Neumann entropy. Also worth adding a test specifically to check the type of the measurement process and its methods respectively.

**Review by Jaybsoni — COMMENTED** [review]
Looks good to me, 2 small comments, happy to approve otherwise!

**Comment by Jaybsoni** [comment]
image

It seems on the device level we have access to the backend we want to cast too. How does that get transferred here when you use `qml.math.vn_entropy`?

File: pennylane/measurements/vn_entropy.py (hunk 1 above)
Comment (AlbertMitjans) [reply pennylane/measurements/vn_entropy.py:75]: Changed to "complex128". Added interface checks.

File: tests/measurements/test_vn_entropy.py (hunk 2 above)
Comment (AlbertMitjans) [reply tests/measurements/test_vn_entropy.py:20]: Done!

**Comment by AlbertMitjans** [comment]
> It seems on the device level we have access to the backend we want to cast too. How does that get transferred here when you use `qml.math.vn_entropy`?

The dtype info is transferred through the `state` argument. Now it is tested too.

**Review by Jaybsoni — APPROVED** [review]
Awesome 💯!

File: pennylane/measurements/vn_entropy.py [hunk 3]
```diff
@@ -61,4 +64,13 @@ def circuit_entropy(x):
...
-    return MeasurementProcess(VnEntropy, wires=wires, log_base=log_base)
+    return _VnEntropy(VnEntropy, wires=wires, log_base=log_base)
+
+
+class _VnEntropy(StateMeasurement):
+    """Measurement process that returns the Von Neumann entropy."""
+
+    def process_state(self, state: Sequence[complex], wires: Wires):
+        return qml.math.vn_entropy(
+            state, indices=self.wires, c_dtype="complex128", base=self.log_base
```
Comment (eddddddy) [inline pennylane/measurements/vn_entropy.py:75]: Again, we should preserve the dtype of the state.

File: pennylane/measurements/vn_entropy.py [hunk 4]
```diff
@@ -61,4 +64,13 @@ def circuit_entropy(x):
     .. seealso:: :func:`pennylane.qinfo.transforms.vn_entropy` and :func:`pennylane.math.vn_entropy`
     """
     wires = Wires(wires)
-    return MeasurementProcess(VnEntropy, wires=wires, log_base=log_base)
+    return _VnEntropy(VnEntropy, wires=wires, log_base=log_base)
+
+
+class _VnEntropy(StateMeasurement):
+    """Measurement process that returns the Von Neumann entropy."""
```
Comment (eddddddy) [inline pennylane/measurements/vn_entropy.py:71]: So the only measurement processes that use the `log_base` argument are `vn_entropy` and `mutual_info`. I think we should i) remove the `log_base` argument from the `__init__` method of `MeasurementProcess` and ii) add `__init__` methods for the `MutualInfo` and `VnEntropy` classes to include this argument.

This will help declutter the code in the base class.

File: pennylane/measurements/vn_entropy.py (hunk 3 above)
Comment (eddddddy) [reply pennylane/measurements/vn_entropy.py:75]: And add a test using a different `c_dtype` for the device.

Comment (AlbertMitjans) [reply pennylane/measurements/vn_entropy.py]: Changed to `state.dtype`.

I realised that right now `DefaultQubit.state.dtype` is always `np.complex128` (even if you change the `c_dtype` of the device). When integrating this functionality with the device, we could either cast the state to the `self.c_dtype` before passing it into the measurement, or cast the result of the measurement to `self.c_dtype`.

What do you think?

File: pennylane/measurements/vn_entropy.py (hunk 4 above)
Comment (AlbertMitjans) [reply pennylane/measurements/vn_entropy.py:73]: Nice catch! Moved `log_base` to children classes

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -64,4 +67,3 @@ Utkarsh Azad
 Soran Jahangiri
 Christina Lee
 Albert Mitjans Coma
```
Comment (AlbertMitjans) [inline doc/releases/changelog-dev.md:69]:
```suggestion
Albert Mitjans Coma

```

File: pennylane/measurements/measurements.py [hunk 6]
```diff
@@ -127,27 +127,23 @@ class MeasurementProcess:
...
         self.return_type = return_type
         self.obs = obs
         self.id = id
-        self.log_base = log_base
 
         if wires is not None and obs is not None:
             raise ValueError("Cannot set the wires if an observable is provided.")
 
-        self._wires = wires or Wires([])
+        self._wires = wires
```
Comment (AlbertMitjans) [inline pennylane/measurements/measurements.py:146]: We keep `self._wires` as the input wires. This will help us simplify the `__copy__` method.

When calling `MeasurementProcess.wires` (public one) we return `Wires([])` if `self._wires` is `None`.

File: tests/measurements/test_vn_entropy.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    def test_queue(self):
+        """Test that the right measurement class is queued."""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev)
+        def circuit():
+            return qml.vn_entropy(wires=0, log_base=2)
+
+        _ = circuit()
```
Comment (eddddddy) [inline tests/measurements/test_vn_entropy.py:57]:
```suggestion
        circuit()
```

File: pennylane/measurements/vn_entropy.py (hunk 3 above)
Comment (eddddddy) [reply pennylane/measurements/vn_entropy.py]: That's strange 🤔 . We should be casting in the [`device.state`](https://github.com/PennyLaneAI/pennylane/blob/f1bf5441a5b704da412a01a5b26cdf42279fee3e/pennylane/devices/default_qubit.py#L663) property, but we don't. We already do the casting in [`device.density_matrix`](https://github.com/PennyLaneAI/pennylane/blob/f1bf5441a5b704da412a01a5b26cdf42279fee3e/pennylane/_qubit_device.py#L1172) so there shouldn't be any issues.

I think we can leave this as `state.dtype` for now, and change the `device.state` property in a new PR to do the casting.
